espanso: 2.1.8 -> 2.2-unstable-2024-05-14

malt3 opened the pull request:
## Description of changes

Upgrade espanso to a pre-release version to fix rendering issue on wayland.
This PR is on the `dev` branch, but not yet in a release: https://github.com/espanso/espanso/pull/1901
Without it, the tool is broken under wayland.

I am unable to test this under darwin.

Closes #286384, closes #291484.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

malt3 committed:
espanso: 2.1.8 -> 2.2.1

malt3 committed:
espanso: 2.2.1 -> 2.2-unstable-2024-05-14

n8henrie commented:
I would really love to get my Darwin changes integrated as well, outlined in https://github.com/NixOS/nixpkgs/pull/291484

I can make them as a separate PR.

Most importantly, I think this pre-release should contain https://github.com/espanso/espanso/commit/471c2841977aa650a08a786679adf2697706c69b which fixes a runtime crash-on-launch for aarch64-Darwin (I can confirm soon).

malt3 committed:
espanso: update darwin patches

This commit also removes patches that seem no longer necessary, removes
`--replace` calls that were not working, and uses `--replace-fail` going
forward in keeping with the deprecation of `--replace`.

Co-authored-by: Nathan Henrie <nate@n8henrie.com>

malt3 commented:
> I would really love to get my Darwin changes integrated as well, outlined in #291484

I tried to incorporate your changes here. Please tell me if this is correct!

malt3 committed:
espanso: add n8henrie as maintainer

n8henrie reviewed (approved):
Runs ans works great, looks to have incorporated all the changes I was interested in. Thank you!

katexochen commented on pkgs/applications/office/espanso/default.nix:
```diff
@@ -39,7 +39,7 @@ assert stdenv.isDarwin -> !x11Support;
 rustPlatform.buildRustPackage rec {
   pname = "espanso";
-  version = "2.2.2-pre";
+  version = "2.2-unstable-2024-05-14";
```
I think this line should go into the previous commit, the rest of the changes into the next one.

eclairevoyant commented on pkgs/applications/office/espanso/default.nix:
```diff
@@ -108,47 +106,47 @@ rustPlatform.buildRustPackage rec {
     license = licenses.gpl3Plus;
-    maintainers = with maintainers; [ kimat pyrox0 ];
+    maintainers = with maintainers; [ kimat pyrox0 n8henrie ];
     platforms = platforms.unix;
```
```suggestion
    platforms = platforms.unix;
    broken = stdenv.hostPlatform.system == "x86_64-darwin";
```
since it seems to be broken on that system per CI

malt3 replied:
Thank you for pointing my at the CI failure. It references this symbol which was apparently added in the `apple_sdk_11_0` (or at least after the default sdk chosen for `x86_64-darwin`): [`kCFURLVolumeAvailableCapacityForImportantUsageKey`](https://developer.apple.com/documentation/corefoundation/kcfurlvolumeavailablecapacityforimportantusagekey). I'm now testing if a change in the SDK fixes the issue.

malt3 replied:
I currently see issues like this: 

```
cargo:warning=/nix/store/ck3awwmd2pca7cs50jbghcxfh32k1jc2-apple-framework-ColorSync-11.0.0/Library/Frameworks/ColorSync.framework/Headers/ColorSyncBase.h:96:69: note: expanded from macro 'CS_UNAVAILABLE_EMBEDDED'
  cargo:warning=#define CS_UNAVAILABLE_EMBEDDED API_UNAVAILABLE(ios, tvos, watchos, macCatalyst)
  cargo:warning=                                                                    

```

Will mark it as broken if I don’t find a solution.

n8henrie replied:
I'm also seeing the error trying to compile sys-locale for x86_64-darwin.

I found [this issue](https://github.com/NixOS/nixpkgs/issues/214872) with a similar error fixed in [this PR](https://github.com/NixOS/nixpkgs/pull/215782) by @wegank . Unfortunately a similar change here (dropping `Foundation`, using `callPackage` from the 11.0 SDK, and explicitly inheriting frameworks from the 11.0 SDK) also fails.

Here's another open issue with a similar error: https://github.com/NixOS/nixpkgs/issues/210558 -- again @wegank comes to the rescue (with the suggestion to take dependencies from the 11.0 SDK).

wegank replied:
I was informed a few hours ago (https://github.com/NixOS/nixpkgs/pull/312654#discussion_r1605794718) that the workaround actually stopped working since half a year ago, I have no idea what changed on the Rust side.

malt3 replied:
Marked x86_64-darwin as broken for now. Seems to be the best way forward to fix issues under linux with wayland and aarch64-darwin.

eliandoran commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5135,7 +5135,8 @@ with pkgs;
 
   espanso = callPackage ../applications/office/espanso {
-    inherit (darwin.apple_sdk.frameworks) AppKit Cocoa Foundation IOKit Kernel AVFoundation Carbon QTKit AVKit WebKit;
+    stdenv = if stdenv.isDarwin then darwin.apple_sdk_11_0.stdenv else stdenv;
```
This can be simplified to just replacing `callPackage` with `darwin.apple_sdk_11.0.callPackage`

According to the documentation `doc/stdenv/platform-notes.chapter.md`:

> x86_64-darwin uses the 10.12 SDK by default, but some software is not compatible with that version of the SDK. In that case,
  the 11.0 SDK used by aarch64-darwin is available for use on x86_64-darwin. To use it, reference `apple_sdk_11_0` instead of
  `apple_sdk` in your derivation and use `pkgs.darwin.apple_sdk_11_0.callPackage` instead of `pkgs.callPackage`. On Linux, this will
  have the same effect as `pkgs.callPackage`, so you can use `pkgs.darwin.apple_sdk_11_0.callPackage` regardless of platform.

Nevertheless, this does not fix the build on darwin-x86_64. You seem to have quite a distinct problem. I have found a relatively similar issue here https://github.com/zed-industries/zed/issues/7036 , so it might be related to the clang version. Not sure why this happens on only a few packages, though, as I've upgraded a few packages to SDK11 with no issues.

malt3 replied:
Removed the explicit sdk selection for now since it doesn't fully fix the build. I think this should be done as a follow-up and once a real solution is known.

msanft commented:
Result of `nixpkgs-review pr 311724` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>espanso</li>
    <li>espanso-wayland</li>
  </ul>
</details>

msanft reviewed (approved):
Lovely, thanks
